In [28]:
import pandas as pd

In [29]:
df = pd.read_csv('jokes-ratings.csv')
df.head()

id  user_id  joke_id  Rating
0  31030_110    31030      110   2.750
1  16144_109    16144      109   5.094
2    23098_6    23098        6  -6.438
3   14273_86    14273       86   4.406
4  18419_134    18419      134   9.375

In [30]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1092059 entries, 0 to 1092058
Data columns (total 4 columns):
 #   Column   Non-Null Count    Dtype  
---  ------   --------------    -----  
 0   id       1092059 non-null  object 
 1   user_id  1092059 non-null  int64  
 2   joke_id  1092059 non-null  int64  
 3   Rating   1092059 non-null  float64
dtypes: float64(1), int64(2), object(1)
memory usage: 33.3+ MB


In [31]:
df['user_id'].nunique()

40863

In [33]:
df['joke_id'].nunique()

139

40863 people rated 139 movies and the total number of ratings given are 1092059

In [34]:
df.drop("id",axis=1, inplace=True)

In [35]:
from surprise import Reader, Dataset

In [38]:
reader = Reader()
data = Dataset.load_from_df(df, reader)

In [20]:
data

SVD

In [39]:
from surprise import SVD, KNNBasic

In [40]:
algo = SVD()

In [41]:
from surprise.model_selection import cross_validate

In [42]:
cross_validate(algo, data, measures= ["RMSE"], cv = 5 , verbose = True)

Evaluating RMSE of algorithm SVD on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    4.6634  4.6617  4.6712  4.6519  4.6573  4.6611  0.0064  
Fit time          10.59   11.04   10.73   10.71   10.32   10.68   0.23    
Test time         1.63    1.54    1.70    1.50    2.12    1.70    0.22    


{'test_rmse': array([4.66336046, 4.66167693, 4.67118054, 4.65194181, 4.65733519]),
 'fit_time': (10.591786861419678,
  11.04238224029541,
  10.731384992599487,
  10.713992834091187,
  10.32496190071106),
 'test_time': (1.6294724941253662,
  1.54060697555542,
  1.7026870250701904,
  1.4966235160827637,
  2.115797758102417)}

In [43]:
from surprise import KNNBasic

In [44]:
knn_algo = KNNBasic()

In [45]:
#cross_validate(knn_algo, data, measures= ["RMSE"], cv = 5 , verbose = True)

In [46]:
df_test = pd.read_csv('jokes-test.csv')
df_test.head(10)

id  user_id  joke_id
0    6194_11     6194       11
1    19356_3    19356        3
2   23426_79    23426       79
3    40030_3    40030        3
4  19806_115    19806      115
5   10487_45    10487       45
6    19959_8    19959        8
7   10979_66    10979       66
8    34886_6    34886        6
9   4729_103     4729      103

In [47]:
df_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 537880 entries, 0 to 537879
Data columns (total 3 columns):
 #   Column   Non-Null Count   Dtype 
---  ------   --------------   ----- 
 0   id       537880 non-null  object
 1   user_id  537880 non-null  int64 
 2   joke_id  537880 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 12.3+ MB


In [48]:
df_test.nunique()

id         537880
user_id     40863
joke_id       139
dtype: int64

40863 users voted for 139 jokes. Total of 537880 votes

In [49]:
df_test.drop('id', axis=1, inplace=True)

In [51]:
test_reader = Reader()
test_data = Dataset.load_from_df(df_test, test_reader)

ValueError: not enough values to unpack (expected 3, got 2)